### ISSUE TITLE
[C+ Checklist Needs Completion] [$500] Android - Scan - Second tap on capture button causes error to show up

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#36520]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.41-3
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
Precondition: Camera access is allowed
1. Launch New Expensify app
2. Tap + > Request money
3. Go to Scan
4. Tap on the capture button twice

## Expected Result:
The second tap on the capture button should be disabled and no error will show up

## Actual Result:
The second tap on the capture button is enabled, causing error to show up

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/dc4563ff-7d0e-4157-b8e9-cca754bf634c

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017ba8b79a30323bee</li>
        <li>Upwork Job ID: 1757796267721375744</li>
        <li>Last Price Increase: 2024-02-14</li>
<li>Automatic offers: </li>
<ul>
<li>hoangzinh | Reviewer | 0</li>
</ul></li>
    </ul>
</details>

**Comment by lanitochka17** [comment]
We think that this bug might be related to #wave5 
CC @quinthar

**Comment by agent3bood** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Android - Scan - Second tap on capture button causes error to show up

### What is the root cause of that problem?

In `IOURequestStepScan.capturePhoto` there is an async operation to capture the photo `camera.current.takePhoto().then()`, only after the async operation we do the navigation/error handling.
Between the time of requesting the first photo and getting the results if the user calls the capture button again, the camera is already busy and we get an error.

### What changes do you think we should make in order to solve the problem?

Add a new flag `capturePhotoInProgress` and return early from `capturePhoto` if it is true. Set the variable to true when we start the function and reset it when it is done.

### What alternative solutions did you explore? (Optional)

**Comment by EnderKilledYou** [comment]
So I see these on upwork. Are they taken or finished since they have a proposal? If I apply there am I wasting connects? I can update this one for sure. You can also disable the button or remove the click handler and add it back later.

**Comment by EnderKilledYou** [comment]
Contributor details
Your Expensify account email: kooshesh@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/sterlingkooshesh

**Comment by EnderKilledYou** [comment]
## Proposal
Update the button handler to disable the button to prevent further inputs
### Please re-state the problem that we are trying to solve in this issue.
The button is able to bel clicked multiple times before being prevented
### What is the root cause of that problem?
Not disabling the button via a state on first click
### What changes do you think we should make in order to solve the problem?

[Update the button handler to disable the button to prevent further inputs](https://github.com/Expensify/App/blob/04d7efd051b83a591ed0cbfed563b47df7c89fff/src/pages/iou/request/step/IOURequestStepScan/index.native.js#L295)
On success or error, reenable it.

USe the disabled tag. This is very common practice and is well tested.

**Comment by greg-schroeder** [comment]
Still waiting on new proposals given @hoangzinh's comment above

**Comment by agent3bood** [comment]
> @agent3bood @Krishna2323 @EnderKilledYou Thanks for proposals, everyone. I think we found the reason why it shows error but I don't think we found the root cause yet. The capture button is using `Button` component, which is already using `useSingleExecution` hook underneath => Expecting it already supported preventing double click issue
> 
> https://github.com/Expensify/App/blob/9843f6641f65d11e125c4325df1f95feae9d1259/src/components/Pressable/GenericPressable/BaseGenericPressable.tsx#L44
> 
> We need to find out why it doesn't work in this case.

The issue is because of the async function `.then()`, so `useSingleExecution` will not work here.
Let me explain 

```
// 1. `useSingleExecution` will make sure this is called just once 
capturePhoto() {
  // ... code
  
  // 2. this async callback function is added to the event loop to be executed in future tick/time.
  camera.current.takePhoto().then() => {
    // 4. some time later, the event loop will execute this code, it has nothing to do with `useSingleExecution`, and if the camera is busy it will throw.
  } 

  //
  // ... code
 // 3. function code is completed, `useSingleExecution` job is done here.
}

```



Buy why are we still to click twice?
Because `useSingleExecution` will prevent second execution in the same "tick", but it does not work for async functions, and for that we need a lock/mutex.

**Comment by agent3bood** [comment]
@hoangzinh yes it is possible, I have checked the code, for the native `singleExecution` to work the `action` must return a `Promise`.
But because the `onPress` handler is passed on multiple levels it is not returning the value of the original.
To make it work first we do `return camera.current.takePhoto(.....` 
Then on every child component that wraps the handler it has to return the result as well.
For example [this](https://github.com/Expensify/App/blob/211884b974fbf682b419ae8b50b0aa2235e3e1af/src/components/Pressable/GenericPressable/BaseGenericPressable.tsx#L112) must do 
```
const res = onPress(event);
Accessibility.moveAccessibilityFocus(nextFocusRef);
return res;

```


Should I update my proposal with the new fix?

**Comment by agent3bood** [comment]
@hoangzinh proposal updates, web is not affected by this issue.

**Comment by agent3bood** [comment]
@hoangzinh updated.

> Fix all components that wrap capturePhoto to return a Promise
I added two places where it need to be updated, `GenericPressable`.

**Comment by agent3bood** [comment]
Applied for the job on Upwork
PR will be ready on Friday.

**Comment by agent3bood** [comment]
Update

@hoangzinh 
There is still work to be done, just doing `return camera.current.takePhoto` is not enough because there are two other async operations to be awaited.

1. the navigation that is done be `navigateToConfirmationStep()` need to be awaited.
2. after the navigation we also need to await end of animation transition.
Until the two are done the button needs to be disabled.

My first solution was using `useWaitForNavigation()` and then await `CONST.ANIMATED_TRANSITION`, but I want to see if there are similar patterns in the app being used.

Will need another day until I can open the PR.

**Comment by agent3bood** [comment]
@hoangzinh what I am saying is there are more async operation going on; originally I thought it was only `takePhoto`; am working on finding the best solution technically and for the UX as well.

What I suggest is in addition to the original proposal. On the `navigateToConfirmationStep` I lock any interaction with the screen once the photo is taken. The navigation will be called and transition to next step.

**Comment by agent3bood** [comment]
btw, I have sent a proposal request in Upwork and waiting for it to be accepted.

**Comment by greg-schroeder** [comment]
This was deployed to staging, but it seems the behavio we intended to fix is still happening per https://github.com/Expensify/App/pull/37182#issuecomment-1973388484

Does this need to be adjusted/reverted before it goes to prod?

**Comment by greg-schroeder** [comment]
@techievivek were you able to weigh in on [this](https://github.com/Expensify/App/issues/36520#issuecomment-1974233547)?

**Comment by greg-schroeder** [comment]
Will process payments this eve

**Comment by techievivek** [comment]
@hoangzinh 

> the current fix still works for double-tripple taps. So in my opinion, we can deliver the current fix. But still discuss with @techievivek if it's worth handling multiple taps in the second PR here https://github.com/Expensify/App/pull/37182#issuecomment-1974018020

I think it will be great if we can fix this completely, even when you keep tapping till the page transitions.

**Comment by agent3bood** [comment]
> @agent3bood are you able to make another PR for this https://github.com/Expensify/App/issues/36520#issuecomment-1990246780? Please

Yes I can do that, will open PR in a day or two.

**Comment by greg-schroeder** [comment]
PR is merged. Awaiting deploy to staging -> prod

**Comment by greg-schroeder** [comment]
Payment summary:

@agent3bood - $500 - C
@hoangzinh - $500 - C+

**Comment by greg-schroeder** [comment]
The upwork issue closed for inactivity, creating a new one via https://github.com/Expensify/App/issues/40222

Offers being sent from that one

**Comment by techievivek** [comment]
Regression test looks good to me. @greg-schroeder Can we get this added? Thanks.
